### PR TITLE
Visual

### PR DESCRIPTION
**Pull request by artur-shaik**
Added option to webview for selection enabled caret mode.
In status bar checking value of this option to identificate about it.

Added bindings: <Space> for toggle selection mode, <Ctrl+Space> drop
selection and keep selection mode enabled.

In webview added javascript snippet to position caret at top of the
viewport after caret enabling. This code mostly was taken from cVim sources. (Maybe should be moved to somewhere better place?)

### CODE REVIEW

File: qutebrowser/keyinput/basekeyparser.py [hunk 1]
```diff
@@ -180,6 +180,7 @@ def _handle_single_key(self, e):
 
         count, cmd_input = self._split_count()
 
+        print(count, cmd_input)
```
Comment (The-Compiler) [inline qutebrowser/keyinput/basekeyparser.py]: That should probably be removed now, or replaced by a debug log command :wink:

File: qutebrowser/keyinput/modeparsers.py [hunk 2]
```diff
@@ -228,19 +228,8 @@ def __init__(self, win_id, parent=None):
         super().__init__(win_id, parent, supports_count=True,
                          supports_chains=True)
         self.read_config('caret')
+        self._supports_count = False
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: Actually having support for counts (like `5w`) would be very useful - for most commands you could probably just do something like:

```python
for _i in range (count):
    webview.triggerPageAction(act)
````

File: qutebrowser/mainwindow/statusbar/bar.py [hunk 3]
```diff
@@ -269,12 +269,12 @@ def caret_active(self):
         return self._caret_active
 
     @pyqtProperty(bool)
-    def visual_active(self):
-        """Getter for self.visual_active, so it can be used as Qt property."""
-        return self._visual_active
+    def caret_selection_active(self):
+        """Getter for self.caret_active, so it can be used as Qt property."""
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/bar.py]: Minor nitpick: That should be `self.caret_selection_active` in the comment.

**Comment by The-Compiler** [comment]
The javascript snippet is a good workaround! :smile:

I suggest moving it to its own file (e.g. at `qutebrowser/javascript/visual_position_caret.js` or so), and then you can do `evaluateJavaScript(utils.read_file('javascript/visual_position_caret.js'))`. Please also add a license header (from cVim I guess) to avoid issues when packaging (e.g. the debian packaging guidelines want a license header in every file).

**Comment by artur-shaik** [comment]
Oops... I was a bit rushed.

About counts I think yes, it can be usefull, but have you any ideas how to use '0' for jumping to start of line?

**Comment by The-Compiler** [comment]
I just looked at the cVim sources, and I think there's enough changed so we don't have to explicitely put their (MIT) license in. I suggest adding a comment with the usual GPL header (like in the other files), and then a comment mentioning some code was taken from https://github.com/1995eaton/chromium-vim/blob/master/content_scripts/visual.js which also makes it easier to find in the future.

Regarding binding of <kbd>0</kbd>, that's currently not possible as it's handled as count, but I think it shouldn't be too hard to edit some stuff in `qutebrowser/keyinput` to allow it to be bound - the first place you'd have to look at is probably `qutebrowser/keyinput/basekeyparser.py:_split_count()` - maybe checking if `count` is `0`, and if so, adding it to `cmd_input` instead of `count` will already be enough.

**Comment by artur-shaik** [comment]
Ok. Thanks for info, everything noticed, will look at it... :smile:

**Comment by artur-shaik** [comment]
Ok, made some changes. I hope nothing was forgotten. :smile:

**Comment by The-Compiler** [comment]
Thanks and sorry for the delay! Merging this into the `visual` branch now, then I'll do a final code review and test and merge that to master.

One small change: I'll change `statusbar.bg.caret_selection` to `statusbar.bg.caret-selection` as qutebrowser uses `-` instead of `_` in all other settings too.

**Comment by artur-shaik** [comment]
Very nice. Thank you.

**Comment by The-Compiler** [comment]
I noticed in d594798db8ecb635ecb7b9ce89d0eb22c27ecc57 you removed the `v` keybinding - was this intentional?

As people in the IRC channel seemed to like it, I think `v` should be readded for `toggle-selection`. What do you think?

I also did various cleanups to the sources, added [eslint](http://www.eslint.org/) as a linter and wrote some tests (#668).

**Comment by artur-shaik** [comment]
`v` key was deleted because of your suggestion. And you know, I found that I always try to enable selection mode with `v` key. So, may be people are right, it better to make `v` key for `toggle-selection`.

**Comment by The-Compiler** [comment]
> `v` key was deleted because of your suggestion. And you know, I found that I always try to enable selection mode with `v` key. So, may be people are right, it better to make `v` key for `toggle-selection`.

Okay, I'll re-add it then. Sorry for the confusion!

What about handling enter similiar to what tmux does? I'd like this workflow:

* Press `v` to go to caret mode.
* Press `Space` to start a selection.
* Select stuff.
* Press `Enter` to go to back to normal mode and yank the selected text.

I can add that if you agree.

Also, thanks for your patience! As you can see, I'm a perfectionist, so I'm sorry this takes some time until it's merged. After those changes, I have one or two places in the code I want to take a closer look at, and after that's done I'll finally merge. :smile:

**Comment by artur-shaik** [comment]
He he, this is nice to be perfectionist. And don't worry about that, I'm fine with it.

For me better workflow is:
- Press `c` to go to caret;
- Press `v` to start selection (As I said, I always instinctively try to toggle selection with `v`);
- Select stuff;
- Press `y` (Enter for me too far away :smile: , and not common) to yank and go to normal mode (Now it keeps selection after yank, but I found that I don't need to keep selection in most time, so it better to go to normal, and drop selection).

**Comment by The-Compiler** [comment]
Okay! I changed `:yank-selected` to leave the mode by default, unless `--stay` is given, so now `y`/`Y` and `Enter` leave the mode.

I also fixed a bug which caused `<Space>` not to be bound correctly (only `v`) and cherry-picked it to the visual branch.

Can you please delete the `[caret]` section from your `keys.conf`, restart and confirm the workflow feels right to you? :smile:

**Comment by artur-shaik** [comment]
Yeah, everything is nice as I want it. :smile:

**Comment by The-Compiler** [comment]
It's now all merged, thank you again! :+1:

**Comment by artur-shaik** [comment]
Nice, thank you too! :smile:

**Comment by The-Compiler** [comment]
Have you seen #675? I wonder if that happened for you during testing as well...

And if you feel like and have some time, I'd still be glad about a spatial navigation mode :smile:

**Comment by artur-shaik** [comment]
Yeah, I saw that. Will try to reproduce it little bit later.

About spatial, of cause I'll be glad to implement. But I need to
understand exactly, what do we want from that.

**Comment by demure** [comment]
So, I recently found this mode, but find it a very un-vim like. To select text in Caret Selection Mode, I must hold down shift. Might we consider adding the three vim visual block modes? These could 'lock' you into selecting without holding shift, and line/block modes are nice.

I also notice I can't use "/", "f", "F", "t", "T", or other vim fast navigation tricks in this mode. I think adding such bindings would make our caret mode quiet epic.

**Comment by The-Compiler** [comment]
@demure Is this on QtWebEngine, where caret mode isn't implemented at all yet? :wink:

**Comment by demure** [comment]
Ah, sorry. When I saw that #626 linked here, I assumed this was the current ticket to comment on.

**Pull request merged**